jctools-benchmarks/src/main/java/org/jctools/jmh/latency/MpqBurstCost.java

```diff
@@ -28,130 +34,253 @@
 @SuppressWarnings("serial")
 public class MpqBurstCost
 {
-    static final Go GO = new Go();
+    private static final long DELAY_PRODUCER = Long.getLong("delay.p", 0L);
+    private static final long DELAY_CONSUMER = Long.getLong("delay.c", 0L);
     @Param( {"SpscArrayQueue", "MpscArrayQueue", "SpmcArrayQueue", "MpmcArrayQueue"})
     String qType;
     @Param( {"100"})
     int burstSize;
     @Param("1")
     int consumerCount;
-    MessagePassingQueue<AbstractEvent> q;
-    private Thread[] consumerThreads;
-    private Consumer consumer;
+    @Param("true")
+    boolean warmup;
+    @Param(value = {"132000"})
+    String qCapacity;
+    MessagePassingQueue<Event> q;
+    private ExecutorService consumerExecutor;
+    private Consumer[] consumers;
+    private CountDownLatch stopped;
 
     @Setup(Level.Trial)
     public void setupQueueAndConsumers()
     {
-        q = MessagePassingQueueByTypeFactory.createQueue(qType, 128);
+        if (warmup)
+        {
+            q = MessagePassingQueueByTypeFactory.createQueue(qType, 128);
+
+            final Event event = new Event();
 
-        // stretch the queue to the limit, working through resizing and full
-        for (int i = 0; i < 128 + 100; i++)
+            // stretch the queue to the limit, working through resizing and full
+            for (int i = 0; i < 128 + 100; i++)
+            {
+                q.offer(event);
+            }
+            for (int i = 0; i < 128 + 100; i++)
+            {
+                q.poll();
+            }
+            // make sure the important common case is exercised
+            for (int i = 0; i < 20000; i++)
+            {
+                q.offer(event);
+                q.poll();
+            }
+        }
+        q = MessagePassingQueueByTypeFactory.buildQ(qType, qCapacity);
+        consumers = new Consumer[consumerCount];
+        for (int i = 0; i < consumerCount; i++)
         {
-            q.offer(GO);
+            consumers[i] = new Consumer(q, i);
         }
-        for (int i = 0; i < 128 + 100; i++)
+        consumerExecutor = Executors.newFixedThreadPool(consumerCount);
+    }
+
+    @Setup(Level.Iteration)
+    public void startConsumers()
+    {
+        stopped = new CountDownLatch(consumerCount);
+        final int consumerCount = this.consumerCount;
+        for (int i = 0; i < consumerCount; i++)
         {
-            q.poll();
+            consumers[i].isRunning = true;
+            consumers[i].stopped = stopped;
         }
-        // make sure the important common case is exercised
-        for (int i = 0; i < 20000; i++)
+        for (int i = 0; i < consumerCount; i++)
         {
-            q.offer(GO);
-            q.poll();
+            consumerExecutor.execute(consumers[i]);
         }
-        q = MessagePassingQueueByTypeFactory.createQueue(qType, 128 * 1024);
-        consumer = new Consumer(q);
-        consumerThreads = new Thread[consumerCount];
+    }
```

Comment:
should probably have a count down latch to make the init wait for the consumers to "start" otherwise we will have some noise from consumer scheduling delay.